main/lsp/LSPPreprocessor.cc

```diff
@@ -81,9 +82,34 @@ void LSPPreprocessor::mergeFileChanges(QueueState &state) {
             requestsMergedCounter++;
         }
     }
-    // Prune history for all edits no longer in queue.
-    ttgs.pruneBefore(earliestVersionInQueue);
     ENFORCE(pendingRequests.size() + requestsMergedCounter == originalSize);
+
+    // Check if we should cancel the slow path.
+    const auto &gs = ttgs.getGlobalState();
+    const auto runningSlowPath = gs.getRunningSlowPath();
+    if (runningSlowPath.has_value()) {
+        const auto &[committed, end] = runningSlowPath.value();
+        // Prune history for all messages no longer in queue or being processed by typechecking thread.
+        ttgs.pruneBefore(min(committed, earliestActiveEditVersion));
+        for (auto it = pendingRequests.begin(); it != pendingRequests.end(); it++) {
+            const auto &msg = *it;
+            if (msg->isNotification() && msg->method() == LSPMethod::SorbetWorkspaceEdit) {
+                auto &params = get<unique_ptr<SorbetWorkspaceEditParams>>(msg->asNotification().params);
+                auto combinedUpdates = ttgs.getCombinedUpdates(committed + 1, params->updates.versionEnd);
+                if (combinedUpdates.canTakeFastPath && gs.tryCancelSlowPath(params->updates.versionEnd)) {
+                    logger->error("[PREPROCESSOR] Canceling typechecking!");
```

Comment:
TODO: Remove these / log them properly to debug. These were useful during development.

Reply:
there's `logger->trace(` in case it's helpful